modules/ml/src/main/java/org/apache/ignite/ml/trainers/TrainerTransformers.java

```diff
@@ -0,0 +1,373 @@
+                    datasetBuilder,
+                    ensembleSize,
+                    subsampleRatio,
+                    featureVectorSize,
+                    maxFeaturesCntPerMdl,
+                    featureExtractor,
+                    aggregator,
+                    environment);
+            }
+        };
+    }
+
+    /**
+     * This method accepts function which for given dataset builder and index of model in ensemble generates
+     * task of training this model.
+     *
+     * @param trainingTaskGenerator Training test generator.
+     * @param datasetBuilder Dataset builder.
+     * @param ensembleSize Size of ensemble.
+     * @param subsampleRatio Ratio (subsample size) / (initial dataset size).
+     * @param aggregator Aggregator of models.
+     * @param environment Environment.
+     * @param <K> Type of keys in dataset builder.
+     * @param <V> Type of values in dataset builder.
+     * @param <M> Type of model.
+     * @return Composition of models trained on bagged dataset.
+     */
+    private static <K, V, M extends Model<Vector, Double>> ModelsComposition runOnEnsemble(
+        IgniteTriFunction<DatasetBuilder<K, V>, Integer, IgniteBiFunction<K, V, Vector>, IgniteSupplier<M>> trainingTaskGenerator,
+        DatasetBuilder<K, V> datasetBuilder,
+        int ensembleSize,
+        double subsampleRatio,
+        int featuresVectorSize,
+        int maximumFeaturesCntPerMdl,
+        IgniteBiFunction<K, V, Vector> extractor,
+        PredictionsAggregator aggregator,
+        LearningEnvironment environment) {
+
+        MLLogger log = environment.logger(datasetBuilder.getClass());
+        log.log(MLLogger.VerboseLevel.LOW, "Start learning.");
+
+        List<int[]> mappings = null;
+        if (featuresVectorSize > 0) {
+            mappings = IntStream.range(0, ensembleSize).mapToObj(
+                modelIdx -> getMapping(
+                    featuresVectorSize,
+                    maximumFeaturesCntPerMdl,
+                    datasetBuilder.upstreamTransformersChain().seed() + modelIdx * MAGIC_PRIME))
+                .collect(Collectors.toList());
+        }
+
+        Long startTs = System.currentTimeMillis();
+
+        datasetBuilder
+            .upstreamTransformersChain()
+            .addUpstreamTransformer(new BaggingUpstreamTransformer<>(subsampleRatio));
+
+        List<IgniteSupplier<M>> tasks = new ArrayList<>();
+        List<IgniteBiFunction<K, V, Vector>> extractors = null;
+        if (mappings != null) {
+            extractors = mappings
+                .stream()
+                .map(m -> wrapExtractor(extractor, m))
+                .collect(Collectors.toList());
+        }
+
+        for (int i = 0; i < ensembleSize; i++) {
+            UpstreamTransformerChain<K, V> newChain = Utils.copy(datasetBuilder.upstreamTransformersChain());
+            DatasetBuilder<K, V> newBuilder = withNewChain(datasetBuilder, newChain);
+            int j = i;
```

Comment:
What was your intention here?

Reply:
To pass `i` into lambda (since it should be final or effectively final).